README.md
# Sign In Button
A Flutter plugin for generating sign in buttons for different social media accounts.

## Getting Started

You must add the library as a dependency to your project.
```yaml
dependencies:
 sign_button: ^2.0.6
```

You should then run `flutter packages get`

Now in your Dart code, you can use:

```dart
import 'package:sign_button/sign_button.dart'
```

## Usage Example

It very simple!

```dart
SignInButton(
  buttonType: ButtonType.google,
  onPressed: () {
   print('click');
  })
```

<img src = "https://github.com/emreesen27/Flutter-Sign-Button/blob/assets/btns.png?raw=true">

ButtonSize
```dart
SignInButton(
  buttonType: ButtonType.google,
  buttonSize: ButtonSize.large, // small(default), medium, large
  onPressed: () {
   print('click');
  })
```
<img src = "https://github.com/emreesen27/Flutter-Sign-Button/blob/assets/size.png?raw=true">

ImagePosition
```dart
SignInButton(
  imagePosition: ImagePosition.left, // left or right
  buttonType: ButtonType.google,
  onPressed: () {
   print('click');
  })
```
<img src = "https://github.com/emreesen27/Flutter-Sign-Button/blob/assets/position.png?raw=true">

Customized Button
```dart
SignInButton(
 buttonType: ButtonType.pinterest,
 imagePosition: ImagePosition.right,
 //[buttonSize] You can also use this in combination with [width]. Increases the font and icon size of the button.
 buttonSize: ButtonSize.large,
 btnTextColor: Colors.grey,
 btnColor: Colors.white,
 width: 140,
 //[width] Use if you change the text value.
 btnText: 'Pinterest',
 onPressed: () {
  print('click');
 })
```

<img src = "https://github.com/emreesen27/Flutter-Sign-Button/blob/assets/custom.png?raw=true">


Disabled Button

```dart
SignInButton(
 buttonType: ButtonType.facebook,
 onPressed: null,
),
```

<img src = "https://github.com/emreesen27/Flutter-Sign-Button/blob/assets/disabled_btn.png?raw=true">


### Mini Button
```dart
SignInButton.mini(
 buttonType: ButtonType.github,
 onPressed: () {},
),
```
<img src = "https://github.com/emreesen27/Flutter-Sign-Button/blob/assets/mini_button.png?raw=true">


### Other properties
- btnText 
- btnColor
- btnTextColor
- elevation
- shape
- width  // You can change the value of width when the text size becomes too small.
- padding  // Padding value is automatically adjusted according to the button size. You can give a value if you want
- customImage // It can be used regardless of button type.

CHANGELOG.md
## [2.0.6] - 12.09.2023

* Image can now be changed

## [2.0.5] - 27.08.2023

* Twitter X button added

## [2.0.4] - 26.05.2023

* Discord buttons added

## [2.0.3] - 12.02.2023

* Android example version upgrade
* Dart 3 ready

## [2.0.2] - 07.11.2021

* Github dark button added.

## [2.0.1] - 17.08.2021

* Instagram and appleDark button added.
* Possibility to disable buttons (thx Florian Laforest)

## [2.0.0] - 21.04.2021

* Null Safety migrated

## [1.0.2] - 0.7.02.2021

* Facebook and google dark buttons added.

## [1.0.1] - 11.12.2020

* Amazon and Quora buttons added.

## [1.0.0] - 10.10.2020

* Mini button option added.

## [0.0.9] - 20.09.2020

* Mail,Yahoo,Reddit buttons added.

## [0.0.8] - 12.09.2020

* Code optimization has been done.
* Readme edited.

## [0.0.7] - 01.09.2020

* Documentation comments edited.

## [0.0.6] - 24.08.2020

* Readme edited.

## [0.0.5] - 24.08.2020

* Padding value has been made customizable.
* Microsoft and youtube buttons added.

## [0.0.4] - 22.08.2020

* Example added.
* Readme edited.

## [0.0.3] - 22.08.2020

* Readme edited.
* Bug fixed.

## [0.0.2] - 21.08.2020

* Readme edited.

## [0.0.1] - 21.08.2020

* Initial build for flutter sign in button library.

test/sign_button_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:sign_button/sign_button.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {

  });
}


lib/create_button.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:sign_button/custom_image.dart';
import 'constants.dart';

// ignore: must_be_immutable
class SignInButton extends StatelessWidget {
  // required
  /// [buttonType] sets the style and icons of the button.
  ButtonType buttonType;

  // required
  /// [onPressed] Send a function to trigger the button.
  VoidCallback? onPressed;

  // not required, default left
  /// [imagePosition] set the position of the icon.(left or right)
  ImagePosition? imagePosition;

  // not required, default 5.0
  /// [elevation] set the button's elevation value.
  double elevation;

  // not required, default small
  /// [buttonSize] set the size of the button. (small medium large)
  ButtonSize buttonSize;

  // not required, Gets value according to buttonType.
  /// [btnColor] Set the background color of the button.
  Color? btnColor;

  // not required, Gets value according to buttonType.
  /// [btnDisabledColor] Set the background color of the disabled button.
  Color? btnDisabledColor;

  // not required, Gets value according to buttonType.
  /// [btnTextColor] set the button's text color.
  Color? btnTextColor;

  // not required, Gets value according to buttonType.
  /// [btnDisabledTextColor] set the disabled button's text color.
  Color? btnDisabledTextColor;

  // not required, Gets value according to buttonType.
  /// [btnText] set the button's text.
  String? btnText;

  // not required, Gets value according to buttonSize.
  /// You can change the value of [width] when the text size becomes too small.
  double? width;

  // not required, Gets value according to buttonSize.
  /// [padding] set the button's padding value.
  double? padding;

  /// [_image] value cannot be assigned.Gets value according to [buttonType].
  Widget? _image;

  /// [customImage] use to assign an image, otherwise [_image] is used
  // not required
  CustomImage? customImage;

  /// [_fontSize] value cannot be assigned.Gets value according to [buttonSize].
  double? _fontSize;

  /// [_imageSize] value cannot be assigned.Gets value according to [buttonSize].
  double? _imageSize;

  // not required, button shape.
  /// [shape] set the button's shape.
  ShapeBorder? shape;

  // not required, button model.
  /// [mini] It automatically takes value according to the selected constructor.
  bool mini;

  SignInButton({
    required this.buttonType,
    required this.onPressed,
    this.imagePosition = ImagePosition.left,
    this.buttonSize = ButtonSize.small,
    this.btnColor,
    this.btnDisabledColor,
    this.btnTextColor,
    this.btnDisabledTextColor,
    this.btnText,
    this.elevation = 5.0,
    this.width,
    this.padding,
    this.shape,
    this.customImage,
  }) : mini = false;

  SignInButton.mini({
    required this.buttonType,
    required this.onPressed,
    this.buttonSize = ButtonSize.small,
    this.btnColor,
    this.btnDisabledColor,
    this.elevation = 5.0,
    this.padding,
    this.customImage,
  }) : mini = true;

  bool get _enabled => onPressed != null;

  bool get _disabled => !_enabled;

  @override
  Widget build(BuildContext context) {
    _setButtonSize();
    _createStyle(context);
    return !mini
        ? MaterialButton(
            color: btnColor,
            disabledColor: btnDisabledColor,
            shape: shape ?? StadiumBorder(),
            onPressed: onPressed,
            elevation: elevation,
            child: Container(
              width: width,
              child: Row(
                mainAxisSize: MainAxisSize.min,
                mainAxisAlignment: imagePosition == ImagePosition.left
                    ? MainAxisAlignment.start
                    : MainAxisAlignment.spaceBetween,
                children: [
                  Padding(
                    padding: EdgeInsets.all(padding!),
                    child:
                        imagePosition == ImagePosition.left ? _image : _text(),
                  ),
                  Padding(
                    padding: EdgeInsets.all(padding!),
                    child:
                        imagePosition == ImagePosition.left ? _text() : _image,
                  ),
                ],
              ),
            ),
          )
        : MaterialButton(
            onPressed: onPressed,
            color: btnColor,
            disabledColor: btnDisabledColor,
            child: _image,
            elevation: elevation,
            padding: EdgeInsets.all(padding!),
            shape: CircleBorder(),
          );
  }

  Widget _text() {
    return Text(
      btnText!,
      style: TextStyle(
        fontSize: _fontSize,
        color: _enabled ? btnTextColor : btnDisabledTextColor,
      ),
    );
  }

  void _setButtonSize() {
    if (buttonSize == ButtonSize.small) {
      padding ??= !mini ? 5.0 : 6.0;
      width ??= 200;
      _fontSize = 15.0;
      _imageSize = !mini ? 24.0 : 30.0;
    } else if (buttonSize == ButtonSize.medium) {
      padding ??= !mini ? 5.5 : 6.5;
      width ??= 220;
      _fontSize = 17.0;
      _imageSize = !mini ? 28.0 : 34.0;
    } else {
      padding ??= !mini ? 6.0 : 7.0;
      width ??= 250;
      _fontSize = 19.0;
      _imageSize = !mini ? 32.0 : 38.0;
    }
  }

  void _createStyle(BuildContext context) {
    btnDisabledColor ??= Theme.of(context).disabledColor.withOpacity(0.12);
    btnDisabledTextColor ??= Theme.of(context).disabledColor.withOpacity(0.38);

    _image = Image.asset(
      customImage?.imagePath ?? 'images/${describeEnum(buttonType)}.png',
      package: customImage?.imagePath != null
          ? (customImage?.package ?? null)
          : 'sign_button',
      width: _imageSize,
      height: _imageSize,
    );

    if (_disabled) {
      _image = ColorFiltered(
        colorFilter: ColorFilter.matrix(<double>[
          0.2126, 0.7152, 0.0722, 0, 0, // R1
          0.2126, 0.7152, 0.0722, 0, 0, // R2
          0.2126, 0.7152, 0.0722, 0, 0, // R3
          0, 0, 0, 1, 0, // R4
        ]),
        child: _image,
      );
    }

    switch (buttonType) {
      case ButtonType.facebook:
        btnText ??= 'Sign in with Facebook';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF1877F2);
        break;

      case ButtonType.facebookDark:
        btnText ??= 'Sign in with Facebook';
        btnTextColor ??= Colors.white;
        btnColor ??= Colors.black;
        break;

      case ButtonType.github:
        btnText ??= 'Sign in with Github';
        btnTextColor ??= Colors.black87;
        btnColor ??= Colors.white;
        break;

      case ButtonType.githubDark:
        btnText ??= "Sign in with Github";
        btnTextColor ??= Colors.white70;
        btnColor ??= Color(0xff212121);
        break;

      case ButtonType.pinterest:
        btnText ??= 'Sign in with Pinterest';
        btnTextColor ??= Colors.white;
        btnColor ??= Colors.redAccent;
        break;

      case ButtonType.apple:
        btnText ??= 'Sign in with Apple';
        btnTextColor ??= Colors.black;
        btnColor ??= Color(0xfff7f7f7);
        break;

      case ButtonType.appleDark:
        btnText ??= 'Sign in with Apple';
        btnTextColor ??= Colors.white;
        btnColor ??= Colors.black;
        break;

      case ButtonType.twitter:
        btnText ??= 'Sign in with Twitter';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF1DA1F2);
        break;

      case ButtonType.twitterX:
        btnText ??= 'Sign in with Twitter';
        btnTextColor ??= Colors.white;
        btnColor ??= Colors.black;
        break;

      case ButtonType.linkedin:
        btnText ??= 'Sign in with LinkedIn';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF3282B8);
        break;

      case ButtonType.google:
        btnText ??= 'Sign in with Google';
        btnTextColor ??= Colors.black;
        btnColor ??= Color(0xfff7f7f7);
        break;

      case ButtonType.googleDark:
        btnText ??= 'Sign in with Google';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF4285F4);
        break;

      case ButtonType.youtube:
        btnText ??= 'Sign in with Youtube';
        btnTextColor ??= Colors.black;
        btnColor ??= Colors.white;
        break;

      case ButtonType.microsoft:
        btnText ??= 'Sign in with Microsoft';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF2F2F2F);
        break;

      case ButtonType.tumblr:
        btnText ??= 'Sign in with Tumblr';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF0F4C75);
        break;

      case ButtonType.mail:
        btnText ??= 'Sign in with Mail';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF20639B);
        break;

      case ButtonType.reddit:
        btnText ??= 'Sign in with Reddit';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFFC85417);
        break;

      case ButtonType.yahoo:
        btnText ??= 'Sign in with Yahoo';
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF7C5295);
        break;

      case ButtonType.amazon:
        btnText ??= 'Sign in with Amazon';
        btnTextColor ??= Colors.black87;
        btnColor ??= Colors.white;
        break;

      case ButtonType.quora:
        btnText ??= 'Sign in with Quora';
        btnTextColor ??= Colors.black87;
        btnColor ??= Colors.white;
        break;

      case ButtonType.instagram:
        btnText ??= 'Sign in with Instagram';
        btnTextColor ??= Colors.black87;
        btnColor ??= Colors.white;
        break;

      case ButtonType.discord:
        btnText ??= "Sign in with Discord";
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF5865F2);
        break;

      case ButtonType.discordCircle:
        btnText ??= "Sign in with Discord";
        btnTextColor ??= Colors.white;
        btnColor ??= Color(0xFF5865F2);
        break;

      case ButtonType.custom:
        btnText ??= "Custom Button";
        btnTextColor ??= Colors.black87;
        btnColor ??= Colors.white;
        break;
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:sign_button/sign_button.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Sign in Button Demo',
      theme: ThemeData.light(),
      darkTheme: ThemeData.dark(),
      themeMode: ThemeMode.system,
      home: MyHomePage(title: 'Sign in Button Demo'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, this.title}) : super(key: key);
  final String? title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  String _buttonClick = "click sign in button";

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title!),
      ),
      body: Center(
        child: SingleChildScrollView(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Text(
                '$_buttonClick',
                style: Theme.of(context).textTheme.headlineSmall,
              ),
              SizedBox(
                height: 5.0,
              ),
              SignInButton(
                  buttonType: ButtonType.apple,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "apple";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.appleDark,
                  shape: RoundedRectangleBorder(
                    borderRadius: BorderRadius.circular(5.0),
                  ),
                  onPressed: () {
                    setState(() {
                      _buttonClick = "appleDark";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.facebook,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "facebook";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.facebookDark,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "facebookDark";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.twitter,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "twitter";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.twitterX,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "twitter X";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.github,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "github";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.githubDark,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "githubDark";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.google,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "google";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.googleDark,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "googleDark";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.linkedin,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "linkedIn";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.youtube,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "youtube";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.microsoft,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "microsoft";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.tumblr,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "tumblr";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.pinterest,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "pinterest";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.mail,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "Mail";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.reddit,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "reddit";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.yahoo,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "yahoo";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.amazon,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "amazon";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.quora,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "quora";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.instagram,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "instagram";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.discord,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "discord";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.discordCircle,
                  onPressed: () {
                    setState(() {
                      _buttonClick = "discord";
                    });
                  }),
              // custom button
              SignInButton(
                  buttonType: ButtonType.custom,
                  width: 230,
                  customImage: CustomImage("images/phone.png"),
                  btnText: "Sign in with phone number",
                  onPressed: () {
                    setState(() {
                      _buttonClick = "phone number";
                    });
                  }),
              SignInButton(
                  buttonType: ButtonType.pinterest,
                  imagePosition: ImagePosition.right,
                  //[buttonSize] You can also use this in combination with [width]. Increases the font and icon size of the button.
                  buttonSize: ButtonSize.large,
                  btnTextColor: Colors.grey,
                  btnColor: Colors.white,
                  width: 140,
                  //[width] Use if you change the text value.
                  btnText: 'Pinterest',
                  onPressed: () {
                    setState(() {
                      _buttonClick = "pinterest";
                    });
                  }),
              //disabled button
              SignInButton(
                  buttonType: ButtonType.yahoo,
                  //btnDisabledColor: Colors.grey,
                  //btnDisabledTextColor: Colors.grey[700],
                  onPressed: null),
              SignInButton.mini(
                buttonType: ButtonType.github,
                onPressed: () {},
              ),
              //disabled mini button
              SignInButton.mini(
                buttonType: ButtonType.github,
                //btnDisabledColor: Colors.grey,
                onPressed: null,
              ),
            ],
          ),
        ),
      ),
    );
  }
}
